File: ParteC#/ProyectoProgramacion3/ProyectoProgramacion3Persistencia/DAO/reportes/ReporteVentasDaoImpl.cs
using System.Data;
using System.Data.Common;
using ProyectoProgramacion3Model.Model.reportes;

namespace ProyectoProgramacion3Persistencia.DAO.reportes;

public class ReporteVentasDaoImpl : DefaultBaseDao<ReporteVentas>, IReporteVentasDao
{
    private const int DefaultIdUsuario = 1;

    protected override DbCommand ComandoCrear(DbConnection conn, ReporteVentas modelo)
    {
        var cmd = CrearStoredProcedure(conn, "sp_InsertReporte_Ventas");
        CrearParametro(cmd, "p_fecha_generacion", DateOnlyParam(modelo.fechaGeneracion), DbType.Date);
        CrearParametro(cmd, "p_fecha_inicio_filtro", DateOnlyParam(modelo.fechaInicioFiltro), DbType.Date);
        CrearParametro(cmd, "p_fecha_fin_filtro", DateOnlyParam(modelo.fechaFinFiltro), DbType.Date);
        CrearParametro(cmd, "p_cantidad_registros", modelo.totalVentas);
        CrearParametro(cmd, "p_monto_total", modelo.montoTotalGenerado);
        CrearParametro(cmd, "p_id_usuario", DefaultIdUsuario);
        CrearParametroSalida(cmd, "_id_generado", DbType.Int32);
        return cmd;
    }

    protected override DbCommand ComandoActualizar(DbConnection conn, ReporteVentas modelo)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "UPDATE Reporte_Ventas SET fecha_generacion=@fechaGeneracion, fecha_inicio_filtro=@fechaInicio, fecha_fin_filtro=@fechaFin, cantidad_registros=@cantidad, monto_total=@monto WHERE id_reporte_ventas=@id";
        CrearParametro(cmd, "@fechaGeneracion", DateOnlyParam(modelo.fechaGeneracion), DbType.Date);
        CrearParametro(cmd, "@fechaInicio", DateOnlyParam(modelo.fechaInicioFiltro), DbType.Date);
        CrearParametro(cmd, "@fechaFin", DateOnlyParam(modelo.fechaFinFiltro), DbType.Date);
        CrearParametro(cmd, "@cantidad", modelo.totalVentas);
        CrearParametro(cmd, "@monto", modelo.montoTotalGenerado);
        CrearParametro(cmd, "@id", modelo.idReporte);
        return cmd;
    }

    protected override DbCommand ComandoEliminar(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_DeleteReporte_Ventas");
        CrearParametro(cmd, "p_id_reporte_ventas", id);
        return cmd;
    }

    protected override DbCommand ComandoLeer(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_ListReporteVentasById");
        CrearParametro(cmd, "p_id_reporte_ventas", id);
        return cmd;
    }

    protected override DbCommand ComandoLeerTodos(DbConnection conn)
    {
        return CrearStoredProcedure(conn, "spListReporteVentas");
    }

    protected override ReporteVentas MapearModelo(DbDataReader reader)
    {
        return new ReporteVentas
        {
            idReporte = LeerEntero(reader, "id_reporte_ventas"),
            fechaGeneracion = LeerDateOnly(reader, "fecha_generacion"),
            fechaInicioFiltro = LeerDateOnly(reader, "fecha_inicio_filtro"),
            fechaFinFiltro = LeerDateOnly(reader, "fecha_fin_filtro"),
            totalVentas = LeerEntero(reader, "cantidad_registros"),
            montoTotalGenerado = LeerDouble(reader, "monto_total")
        };
    }
}

File: backend/ParteJAVA/ProyectoPro3Persistencia/src/main/java/pe/edu/pucp/proyectopro3/dao/reportes/ReporteVentasDAOImpl.java
package pe.edu.pucp.proyectopro3.dao.reportes;

import pe.edu.pucp.proyectopro3.dao.DefaultBaseDAO;
import pe.edu.pucp.proyectopro3.modelo.reportes.ReporteVentas;

import java.sql.*;

public class ReporteVentasDAOImpl extends DefaultBaseDAO<ReporteVentas> implements ReporteVentasDAO {
    private static final int DEFAULT_ID_USUARIO = 1;

    @Override
    protected PreparedStatement comandoCrear(Connection conn, ReporteVentas modelo) throws SQLException {
        String sql = "{call sp_InsertReporte_Ventas(?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);

        cmd.setDate(1, new java.sql.Date(modelo.getFechaGeneracion().getTime()));
        cmd.setDate(2, new java.sql.Date(modelo.getFechaInicioFiltro().getTime()));
        cmd.setDate(3, new java.sql.Date(modelo.getFechaFinFiltro().getTime()));
        cmd.setInt(4, modelo.getTotalVentas());
        cmd.setDouble(5, modelo.getMontoTotalGenerado());
        cmd.setInt(6, DEFAULT_ID_USUARIO);
        cmd.registerOutParameter(7, Types.INTEGER);

        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, ReporteVentas modelo) throws SQLException {
        String sql = "UPDATE Reporte_Ventas SET fecha_generacion=?, fecha_inicio_filtro=?, "
                + "fecha_fin_filtro=?, cantidad_registros=?, monto_total=? WHERE id_reporte_ventas=?";
        PreparedStatement cmd = conn.prepareStatement(sql);

        cmd.setDate(1, new java.sql.Date(modelo.getFechaGeneracion().getTime()));
        cmd.setDate(2, new java.sql.Date(modelo.getFechaInicioFiltro().getTime()));
        cmd.setDate(3, new java.sql.Date(modelo.getFechaFinFiltro().getTime()));
        cmd.setInt(4, modelo.getTotalVentas());
        cmd.setDouble(5, modelo.getMontoTotalGenerado());
        cmd.setInt(6, modelo.getIdReporte());

        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_DeleteReporte_Ventas(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_ListReporteVentasById(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "{call spListReporteVentas()}";
        return conn.prepareCall(sql);
    }

    @Override
    protected Integer extraerIdDesdeCallable(CallableStatement cmd) throws SQLException {
        return cmd.getInt(7);
    }

    @Override
    protected Integer extraerIdDesdeGeneratedKeys(ResultSet rs) throws SQLException {
        return rs.getInt(1);
    }

    @Override
    protected ReporteVentas mapearModelo(ResultSet rs) throws SQLException {
        ReporteVentas reporte = new ReporteVentas();

        reporte.setIdReporte(rs.getInt("id_reporte_ventas"));
        reporte.setFechaGeneracion(rs.getDate("fecha_generacion"));
        reporte.setFechaInicioFiltro(rs.getDate("fecha_inicio_filtro"));
        reporte.setFechaFinFiltro(rs.getDate("fecha_fin_filtro"));
        reporte.setTotalVentas(rs.getInt("cantidad_registros"));
        reporte.setMontoTotalGenerado(rs.getDouble("monto_total"));

        return reporte;
    }
}
